MVP support for co-located Jujutsu repositories

ErichDonGubler opened the pull request:
I'm deliberately using a draft PR as a point of collaboration from folks that actually use `jj` and `moz-phab`, and would like them to play better together.

**This PR is now submitted as a patch stack upstream: [bug 1956345](https://bugzilla.mozilla.org/show_bug.cgi?id=1956345)**

Use cases that should work:

- [x] Usable in workspace (thanks @maltejur!)
- [x] Usable with a non-co-located Git repo: punted to #11 
- [x] `moz-phab submit`
	- [x] MVP implementation that WFM on naive usages; not all combinations of `start`, `end`, and `--single` arguments have been tried on my machine, but I _think_ they should Just Work™.
		* Performs much faster than Git and Mercurial, since all commits can be submitted and have their messages updated without needing to change the current worktree. 🥳 
- [ ] `moz-phab patch`
	- [x] MVP implementation
	- [x] Don't change current change when `patch`ing
	- [ ] Preserve author date
	- [x] Create bookmarks like we would create branches in Git
		- [x] Respect `--no-bookmark`? `--no-branch`? Both?
- [x] `moz-phab reorg`: tested, and should be working fully.
- [ ] `moz-phab uplift`: untested
- [ ] Every other subcommand is not guaranteed to work.

# Installation

If you just want to install this without checking out anything, you can use `pip` to install as follows (thanks @mhammond):

```
pip install MozPhab@git+https://github.com/erichdongubler-mozilla/review@refs/pull/1/head
```

If you want to test changes you've made to a local checkout, you can instead use this at the root of the repository:

```
uv tool install --force --editable --refresh .
```

…with your CWD at the root of a local repository.

Once installed, you may wish to ignore auto-updates, so you don't lose Jujutsu support. You can do this by setting `self_auto_update` to `False` in `~/.moz-phab-config`.

ErichDonGubler commented:
Tip: It’s pretty easy to install a dev. copy of `moz-phab`, I've found! You can use `pipx install --editable . --force` with your CWD at the root of a local repository.

maltejur commented on mozphab/jujutsu.py:
```diff
@@ -0,0 +1,433 @@
+            # No commits found to submit
+            return None
+
+        boundary = '"--%s--\n"' % uuid.uuid4().hex
```
```suggestion
        boundary = "--%s--\n" % uuid.uuid4().hex
```

maltejur replied:
The log won't get split properly later if the quotation marks are included directly in `boundary`

ErichDonGubler replied:
Oh, shoot, good catch! 🙂

ErichDonGubler replied:
Fixed with a force-push of the relevant commit to 58d7101d.

maltejur commented on mozphab/jujutsu.py:
```diff
@@ -0,0 +1,433 @@
+                    "commit_id, ",
+                    "conflict, ",
+                    "description, ",
+                    boundary,
```
```suggestion
                    f'"{boundary}"',
```

ErichDonGubler replied:
Note to self: Another piece of <https://github.com/erichdongubler-mozilla/review/pull/1#discussion_r1765315383>.

Fixed with a force-push of the relevant commit to 58d7101d.

maltejur commented on mozphab/jujutsu.py:
```diff
@@ -0,0 +1,433 @@
+            split=False,
+            strip=False,
+        )[: -len(boundary)]
+        changes = log.split(boundary)
```
Currently, the stack gets submitted in the wrong order, so you should probably invert the list of commits at some point.

ErichDonGubler replied:
@maltejur: Uh, I think that this is the same behavior as the Git backend, unless I'm mistaken? 🤔 Which is not unlikely.

maltejur replied:
Uhm, I ran it today and my stack definitely arrived in the wrong order on Phabricator 😅.

![Screenshot_20240918_174721](https://github.com/user-attachments/assets/8265d3f0-7468-475d-b5a7-9ca6c8191334)

I think the point where the Git commits are being reversed is here: https://github.com/erichdongubler-mozilla/review/blob/901957058371ed9c502095d19526ac0b94bec722/mozphab/git.py#L334

ErichDonGubler replied:
Ah, I know what happened, reexamining this code. I had originally written and tested with the  `--reversed` flag in a "bare" `check_output` call to `jj log`, but accidentally broke this by not porting the flag usage over when I refactored later to reuse `self.__cli_log`.

Will fix shortly!

ErichDonGubler replied:
@maltejur: Can you confirm that e60d23ad fixes the issue for you? 🙏🏻

maltejur replied:
Thanks, I can confirm that the commits now show up in inverted order on the `Submitting 7 commits as Work In Progress` confirmation prompt. I sadly don't have a new stack to submit right now to confirm that they indeed also arrive in the correct order on Phabricator now, but I would assume they do.

ErichDonGubler commented:
Some updates in the last force-push:

- Rebased onto 1.6.1 upstream.
- Fixed the end component of an inferred `submit` revision range to be `@-` in the absence of specification or `--single`, rather than the start component.

maltejur commented:
FYI: You can set `self_auto_update` to `False` in `~/.moz-phab-config` to prevent moz-phab from auto-updating itself away from a custom install.

ErichDonGubler commented:
Updates from latest push to 1ad7d9d260ba:

* Rebased onto upstream `1.7.0` tag.
* Renamed `branch` to `bookmark` where it made sense, to migrate to [the deprecation/rename in Jujutsu 0.22](https://github.com/martinvonz/jj/blob/main/CHANGELOG.md#0220---2024-10-02).

ErichDonGubler commented:
Latest force-pushes:

* New tip of d3c33e4 brings in fixes from #4 and #5 ([interdiff link](https://github.com/erichdongubler-mozilla/review/compare/1ad7d9d260ba8bdffd5e7f6c71014d60e1a2969e..9f6dbc3e0d184249920811a2880e0c8e50a5d1fd)).
* New tip of 2ce9df4 squashes the above fixes into the `submit` commit ([interdiff link](https://github.com/erichdongubler-mozilla/review/compare/d3c33e4596878a7829ef622690cc3f6ba429fcfd..2ce9df4a49295d79fb641d9425a4469b4c302ee7)).

ErichDonGubler commented:
Latest force-push to 2c19eb3 rebases this PR on top of 1.8.1 upstream ([interdiff link](https://github.com/erichdongubler-mozilla/review/compare/2ce9df4a49295d79fb641d9425a4469b4c302ee7..2c19eb36e65c6878c585f738275fc76309240334)).

ErichDonGubler commented on mozphab/jujutsu.py:
```diff
@@ -0,0 +1,471 @@
+        if not start_rev:
+            return None
+
+        end_rev = None
+        if is_single:
+            end_rev = start_rev
+        elif (
+            hasattr(self.args, "end_rev")
+            and self.args.end_rev != environment.DEFAULT_END_REV
+        ):
+            end_rev = self.args.end_rev
+        else:
+            end_rev = self.__get_last_mutable_parent()
+
+        self.revset = (start_rev, end_rev)
```
suggestion: @ahal and @Mossop raised in internal Slack that it might be a good idea to change this auto-detection to be include the working-copy commit if it is "ready" (i.e., not empty and with a non-empty description).

ahal replied:
That sounds like a reasonable way to detect it to me!

ErichDonGubler replied:
We might also want to respect [`git.private-commits`](https://martinvonz.github.io/jj/latest/config/#set-of-private-commits) here.

ErichDonGubler commented:
Latest force-push fixes up `moz-phab patch` to actually be usable as an MVP ([interdiff link](https://github.com/erichdongubler-mozilla/review/compare/2c19eb36e65c6878c585f738275fc76309240334..9822f6c1ee38301edbac0a826df4df00f8d5af0c))! You should be able to call `moz-phab patch` from anywhere, including workspaces and subdirectories of your repositories, and it should _Just Work™_. Note on current limitations:

- The current change (`@`) is changed when `patch`ing, since it's currently implemented using `git apply`. Once `jj` upstream exposes a way to do this, we can likely lift this limitation.
- `jj` gained the ability to change a commit author from CLI, so we use that to properly preserve the author for `patch`ed commits. However, the commit author date is still not something `jj` affords from the CLI, so this still isn't preserved when `patch`ing revisions.
- `patch` still does not create a bookmark/branch with Jujutsu, ignoring configuration and CLI options. This just needs additional work to figure out, since Jujutsu doesn't have a notion of "currently checked out branch that should be automatically advanced".

ErichDonGubler commented:
I've changed the developer install instructions to be based on `uv`, rather than `pipx`. It's _so_ much faster (<500 ms) on my local machine for me than `pipx` (>7 sec), it's sorta bonkers.

maltejur commented:
@ErichDonGubler would you mind taking a look at https://github.com/erichdongubler-mozilla/review/pull/6? It's just a very small change.

ErichDonGubler commented:
I’ve rebased the current branch for `jj` support onto 1.9.0. I strongly recommend updating, so you can avoid disruptions with the migration to [`mozilla-firefox/firefox`](https://github.com/mozilla-firefox/firefox)!

ErichDonGubler commented:
This PR and its dependents have been rebased onto 1.11.0. Cheers!

ErichDonGubler commented:
This has now landed in upstream with [bug 1956345](https://bugzilla.mozilla.org/show_bug.cgi?id=1956345). Thanks for the help, everyone! I've migrated concrete issues in the OP to Bugzilla. ❤️
